File: Python/Q11_ClasseAbstrata.py
from abc import ABC, abstractmethod

class Funcionario(ABC):
    @abstractmethod
    def calcularSalario(self):
        pass

class FuncionarioHorista(Funcionario):
    def __init__(self, nome, valorHora):
        self.nome = nome
        self.valorHora = valorHora
    def calcularSalario(self, horas):
        return horas * self.valorHora

class FuncionarioAssalariado(Funcionario):
    def __init__(self, nome, salario):
        self.nome = nome
        self.salario = salario
    def calcularSalario(self):
        return self.salario

funcionario1 = FuncionarioAssalariado("Zé",2000)
funcionario2 = FuncionarioHorista("João",20)

print("Salário Horista: ",funcionario2.calcularSalario(150))
print("Salário Assalariado: ",funcionario1.calcularSalario())

File: go/Q11_ClasseAbstrata/Q11_ClasseAbstrata.go
package main

import (
	"fmt"
)

type Funcionario interface {
	CalcularSalario() float64
}

type FuncionarioHorista struct {
	nome         string
	horasTrabalhadas int
	valorHora    float64
}

func (f FuncionarioHorista) CalcularSalario() float64 {
	return float64(f.horasTrabalhadas) * f.valorHora
}

type FuncionarioAssalariado struct {
	nome  string
	salario float64
}

func (f FuncionarioAssalariado) CalcularSalario() float64 {
	return f.salario
}

func exibirSalario(funcionario Funcionario) {
	fmt.Printf("Salário: R$ %.2f\n", funcionario.CalcularSalario())
}

func main() {
	funcionario1 := FuncionarioHorista{nome:"Zé", horasTrabalhadas: 150, valorHora: 20.00}
	funcionario2 := FuncionarioAssalariado{nome:  "João",salario: 2000.00}

	fmt.Printf("Funcionário: %s\n", funcionario1.nome)
	exibirSalario(funcionario1)

	fmt.Printf("Funcionário: %s\n", funcionario2.nome)
	exibirSalario(funcionario2)
}
